Update: support .eslintrc.cjs (refs eslint/rfcs#43)

evanplaice opened the pull request:
In ES module packages w/ "type": "module" defined treat all .js files as ES modules. Since `eslint` is expecting the config in CommonJS format, it needs to be named w/ a `.cjs` extension.

<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).
-->

**What is the purpose of this pull request? (put an "X" next to item)**

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

For details about the issue, see #12319.
For details about the resolution, see [RFC #43](https://github.com/eslint/rfcs/pull/43)

**What changes did you make? (Give an overview)**

Map the `.cjs` file extension for loading CommonJS configurations (ie `.eslintrc`) in ES module based packages.

**Is there anything you'd like reviewers to focus on?**

After searching through the source, I didn't see any tests that cover this code. Is there an existing test I need to add/update to cover this functionality?

Also, what section of the docs should ES module usage be documented?

kaicataldo commented:
Thanks for the PR. We ask that changes to core go through the RFC process outlined [here](https://github.com/eslint/rfcs). Do you mind doing that?

Incidentally, I don't think we should consider this a bug fix. It would be an enhancement because it would be adding support for an experimental Node feature.

mysticatea commented:
Yeah, this is not a bug fix.

I think too early to merge this feature because ES modules on Node.js is still experimental and it may change due to the large impact to JS ecosystem.

evanplaice commented:
Speaking from experience, any conversation discussing modules + node has a tendency to fall down the rabbit hole. I posted this PR so there's an actual concrete example to look at.

I don't expect this PR to land quickly, if at all. 

@kaicataldo 

>We ask that changes to core go through the RFC process outlined here. Do you mind doing that?

Not at all, it'll take me a bit to gather and record my thoughts
@mysticatea 

Of course. The initial issue said 'bug' so I followed that workflow. I'll fix it once I get a feel for the process.

As far as the time schedule goes -- if nothing major has or will change between now and then -- modules will be officially released (ie unflagged) on 20 October. There's still time to discuss, bike shed, shave a few yaks, and decide the best approach.

evanplaice commented:
Update: As mentioned in [this issue](https://github.com/nodejs/modules/issues/400) it looks like the schedule for unflagging ES modules is going to be pushed a bit.

jkrems commented:
Update: The unflagging PR (https://github.com/nodejs/node/pull/29866) has now landed, an unflagged version of module support in node should be part of the next 13.x release next week.

mysticatea commented:
Hi. Thank you for working on this.

RFC43 has been merged in the last week.
Would you update this PR to follow the RFC?

ljharb commented on docs/user-guide/configuring.md:
```diff
@@ -664,6 +664,7 @@ In each case, the settings in the configuration file override default settings.
 ESLint supports configuration files in several formats:
 
 * **JavaScript** - use `.eslintrc.js` and export an object containing your configuration.
+* **JavaScript (ESM)** - same as `.eslintrc.js` but used with ES module packages
```
if `.eslintrc.js` as ESM can be supported, then should `.eslintrc.mjs` be added to the list as well?

evanplaice replied:
I had the same thought

Currently, ESLint only works with CommonJS-based configs; rendering the need for a `.mjs` extension unnecessary. If support for ESM-based configurations is added in the future then the addition of `.mjs` extension will likely be necessary.

This PR and the preceding RFC was intentionally made with the following assumptions:
- ESLint itself is written in CommonJS
- Javascript configuration files for ESLint will be written in CommonJS format

ljharb replied:
In that case, this docs line seems misleading, since eslint does *not* support "JavaScript (ESM)"-based configuration?

TrejGun replied:
it would be good to see mjs here too

kaicataldo replied:
@TrejGun As discussed above, this is currently not possible. Please follow along on [this RFC](https://github.com/eslint/rfcs/pull/40) to see discussion around how ESLint is trying to move towards being able to support ESM configuration files and plugins.

kaicataldo commented on docs/user-guide/configuring.md:
```diff
@@ -664,6 +664,7 @@ In each case, the settings in the configuration file override default settings.
 ESLint supports configuration files in several formats:
 
 * **JavaScript** - use `.eslintrc.js` and export an object containing your configuration.
+* **JavaScript (ESM)** - same as `.eslintrc.js` but used with ES module packages
```
Should this be `.eslintrc.cjs`? I agree with @ljharb that we could clarify this. Suggestion:

```suggestion
* **JavaScript (ESM)** - use `.eslintrc.cjs` when running ESLint in JavaScript packages that specify `"type":"module"` in their `package.json`. Note that ESLint does not support ESM configuration at this time.
```

We could then get rid of the note below.

evanplaice replied:
Assuming this is a suitable fix, this change was made and the note below was removed in 647feca

evanplaice commented:
I added test cases to:
- "'loadFile(filePath, options)' method should load a config file."
- "'loadInDirectory(directoryPath, options)' method should load the config file of a directory."
- "'extends' property should handle the content of extended configs properly."

And added the test:
- "should load information from a JavaScript file with a .cjs extension"

As well as the fixture:
- /eslint/tests/fixtures/config-file/cjs/.eslintrc.cjs

Is that sufficient? AFAIK, these are the tests that cover config loading. The rest of the tests appear to dig into specific JavaScript-configuration behaviors which should be no different between `.js` vs `.cjs`.

evanplaice commented:
The `CI / Browser Test (pull_request)` is failing [across the board for all PRs](https://github.com/eslint/eslint/actions).

Should I squash + rebase after it's fixed or just leave it?

kaicataldo commented:
`CI / Browser Test` is failing due to https://github.com/eslint/eslint/pull/12662 and unrelated to this PR. We're trying to decide on a fix. We shouldn't hold up merging this for that.

evanplaice commented:
@kaicataldo I'm happy to help. Big thanks to you and everybody else who was involved in this for your patience.

kaicataldo commented:
Thanks for contributing to ESLint!
